### DISCUSSION TITLE
Support for variable width fonts.

### DISCUSSION THREAD

**Discussion by MShark67** [pyscripter/SynEdit#76] [Ideas]
Recently I noticed that SynEdit correctly handles mixing English characters and Korean characters on a line.  Korean characters are wider than English characters and SynEdit handles it perfectly as far as my testing goes.  It got me thinking about variable width character fonts (say Arial.)  Of course, testing (by disabling the fixed width font check) shows that it definitely doesn't work as the cursor position isn't correctly handled.  Any thoughts on if it's feasible or even makes sense to handle non-fixed width fonts?  I'm making this just a discussion as it's not really even a feature request... just something I'd been thinking about.

**Comment by pyscripter** [comment]
Probably all it would take to support variable width fonts is to remove the assumptions of fixed width in:

- TCustomSynEdit.PixelsToColumn
- TCustomSynEdit.ColumnToPixels
- TCustomSynEdit.TextWidth

These assumptions are there to speed up these routines for the common cases (spaces + letters).

**Comment by MShark67** [comment]
Thank you!   Making those changes seems to work and performance changes aren't obvious (though I've only done limited testing.)  I'll do more testing, but I wonder if this would be a useful option to add.  There would just need to be a way to not restrict the font selection to fixed width (maybe a hard option or a define.)  Obviously, we wouldn't want to hurt performance for the people using fixed width fonts (which, after all, is the main focus of SynEdit.)

**Reply by pyscripter** [reply]
It may not be as simple to fully support variable width fonts.   Doing a quick search for CharWidth gives 81 matches in 6 files.   The files include WordWrap.pas and SynEditScrollBars.pas.   Each of these 81 matches would need to be reviewed carefully.   I am reluctant to devote time on this, since I cannot readily see the use cases.

**Reply by MShark67** [reply]
100%.  I totally get it.  I may look at this more closely just because I have a possible use case.  If so, I'll try to make it as clean and segmented as possible.  I appreciate the feedback!
